README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# pksensi <img src="man/figures/logo.png" height="200px" align="right" />

<!-- badges: start -->

[![lifecycle](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Dependencies](https://tinyverse.netlify.com/badge/pksensi)](https://cran.r-project.org/package=pksensi)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/nanhung/pksensi?branch=master&svg=true)](https://ci.appveyor.com/project/nanhung/pksensi)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version-last-release/pksensi)](https://cran.r-project.org/package=pksensi)
[![CRAN\_Download\_Badge](http://cranlogs.r-pkg.org/badges/pksensi)](https://cran.r-project.org/package=pksensi)
[![Total
Downloads](https://cranlogs.r-pkg.org/badges/grand-total/pksensi?color=orange)](https://CRAN.R-project.org/package=pksensi)
[![Codecov test
coverage](https://codecov.io/gh/nanhung/pksensi/branch/master/graph/badge.svg)](https://codecov.io/gh/nanhung/pksensi?branch=master)
[![R-CMD-check](https://github.com/nanhung/pksensi/workflows/R-CMD-check/badge.svg)](https://github.com/nanhung/pksensi/actions)
<!-- badges: end -->

**pksensi** implements the global sensitivity analysis workflow to
investigate the parameter uncertainty and sensitivity in physiologically
based kinetic (PK) models, especially the physiologically based
pharmacokinetic/toxicokinetic model with multivariate outputs. The
package also provides some functions to check the convergence and
sensitivity of model parameters.

Through **pksensi**, you can:

-   Run sensitivity analysis for PK models in R with script that were
    written in C or GNU MCSim.

-   Decision support: The output results and visualization tools can be
    used to easily determine which parameters have “non-influential”
    effects on the model output and can be fixed in following model
    calibration.

## Installation

You can install the released version of **pksensi** from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("pksensi")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("nanhung/pksensi")
```

-   This package includes a function that can help you install GNU MCSim
    more easily through the function `mcsim_install()`.

-   All updated details can be found in
    [NEWS.md](https://github.com/nanhung/pksensi/blob/master/NEWS.md).

-   **NOTE:** Windows users need to install
    [Rtools40](https://cran.r-project.org/bin/windows/Rtools/) to
    compile the model code.

## Workflow

![](https://i.ibb.co/tqpDLrk/sensitivity-workflow.png)

**Note:** The parameter correlation (e.g., V<sub>max</sub> and
K<sub>M</sub> in metabolism) might be an issue in the global sensitivity
analysis. If you have experiment data, suggest using small datasets as a
sample in Markov Chain Monte Carlo Simulation. Then, check correlation
before conducting the sensitivity analysis. The issue will try to
address in the future version.

## Example

This is a basic example of applying **pksensi** in one-compartment pbtk
model:

``` r
library(pksensi)
```

### Step 1. Construct 1-cpt pbtk model

``` r
pbtk1cpt <- function(t, state, parameters) {
  with(as.list(c(state, parameters)), {
    dAgutlument = - kgutabs * Agutlument
    dAcompartment = kgutabs * Agutlument - ke * Acompartment
    dAmetabolized = ke * Acompartment
    Ccompartment = Acompartment / vdist * BW;
    list(c(dAgutlument, dAcompartment, dAmetabolized), 
         "Ccompartment" = Ccompartment) 
  })
}
```

### Step 2. Define initial conditions, output time steps and variable

``` r
initState <- c(Agutlument = 10, Acompartment = 0, Ametabolized = 0)
t <- seq(from = 0.01, to = 24.01, by = 1)
outputs <- c("Ccompartment")
```

### Step 3. Generate parameter matrix

#### 3.1. (Optional) Extract parameter value from httk package

``` r
library(httk)
pars1comp <- (parameterize_1comp(chem.name = "acetaminophen"))
#> Human volume of distribution returned in units of L/kg BW.
```

#### 3.2. Set parameter distributions

``` r
q <- c("qunif", "qunif", "qunif", "qnorm")
q.arg <- list(list(min = pars1comp$Vdist / 2, max = pars1comp$Vdist * 2),
              list(min = pars1comp$kelim / 2, max = pars1comp$kelim * 2),
              list(min = pars1comp$kgutabs / 2, max = pars1comp$kgutabs * 2),
              list(mean = pars1comp$BW, sd = 5))
```

#### 3.3. Create parameter matrix

``` r
set.seed(1234)
params <- c("vdist", "ke", "kgutabs", "BW")
x <- rfast99(params, n = 200, q = q, q.arg = q.arg, replicate = 1)
```

### Step 4. Conduct simulation (will take few minutes with more replications)

``` r
out <- solve_fun(x, time = t, func = pbtk1cpt, initState = initState, outnames = outputs)
#> Starting time: 2022-01-13 09:47:44
#> Ending time: 2022-01-13 09:47:52
```

### Step 5. Uncertainty analysis

``` r
pksim(out)  # Use to compare with "real" data (if any)
```

<img src="man/figures/README-unnamed-chunk-8-1.png" width="100%" />

### Step 6. Check and visualize the result of sensitivity analysis

``` r
plot(out)   # Visualize result
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />

``` r
check(out)  # Print result to console
#> 
#> Sensitivity check ( Index > 0.05 )
#> ----------------------------------
#> First order:
#>  vdist ke kgutabs 
#> 
#> Interaction:
#>  vdist ke kgutabs 
#> 
#> Total order:
#>  vdist ke kgutabs 
#> 
#> Unselected factors in total order:
#>  BW 
#> 
#> 
#> Convergence check ( Index > 0.05 )
#> ----------------------------------
#> First order:
#>   
#> 
#> Interaction:
#>   
#> 
#> Total order:
#> 
```

## Citation


    To cite pksensi in publications use:

      Hsieh, N-H., Reisfeld B., and Chiu W.A., (2020). pksensi: An R
      package to apply global sensitivity analysis in physiologically based
      kinetic modeling SoftwareX, 12, 100609.
      https://doi.org/10.1016/j.softx.2020.100609

    A BibTeX entry for LaTeX users is

      @Article{,
        title = {{pksensi}: An R package to apply global sensitivity analysis in physiologically based kinetic modeling},
        author = {Nan-Hung Hsieh and Brad Reisfeld and Weihsueh A. Chiu},
        journal = {SoftwareX},
        year = {2020},
        volume = {12},
        pages = {100609},
        doi = {10.1016/j.softx.2020.100609},
      }

## Reference

Hsieh NH, Reisfeld B, Bois FY, Chiu WA. [Applying a global sensitivity
analysis workflow to improve the computational efficiencies in
physiologically-based pharmacokinetic
modeling](https://www.frontiersin.org/articles/10.3389/fphar.2018.00588/full).
Frontiers in Pharmacology 2018 Jun; 9:588.

Hsieh NH, Reisfeld B, Chiu WA. [pksensi: An R package to apply global
sensitivity analysis in physiologically based kinetic
modeling](https://doi.org/10.1016/j.softx.2020.100609). SoftwareX 2020
Jul; 12:100609.

Hsieh NH, Bois FY, Tsakalozou E, Ni Z, Yoon M, Sun W, Klein M, Reisfeld
B, Chiu WA. [A Bayesian population physiologically based pharmacokinetic
absorption modeling approach to support generic drug development:
application to bupropion hydrochloride oral dosage
forms](https://doi.org/10.1007/s10928-021-09778-5). Journal of
Pharmacokinetics and Pharmacodynamics 2021 Sep; 22:1-6.


NEWS.md
# pksensi 1.2.3

* Syntax update (check & install)
* Fix bug (check)
* Fix MCSim installation issue (#9)
* Remove dependent: magrittr, reshape2
* New function: `set_rtools40_path()`
* Add feature to run simple simulation (`generate_infile()`)

# pksensi 1.2.2

* Add more flexible parallel method (`solve_mcsim()`)
* Fix error in previous version

# pksensi 1.2.1

* Add `mcsim_pkg()` to download and run portable MCSim
* Change the site location to https://nanhung.github.io
* Improve coverage

# pksensi 1.2.0

* update the route to load example model
* update the default MCSim version to 6.2.0
* Switch the contact mail to d99622005@ntu.edu.tw

# pksensi 1.1.7

* Suggested to use Rtools40
* Added parallel method in `solve_mcsim()`
* Fixed bug in `heat_check()`
* Added message of display info in `heat_check()` function
* Improved `heat_check()` to display the indices based on the setting cut-off

# pksensi 1.1.6

* Improved `heat_check()` to only show influential parameters as default
* Updated `check()` to return the SI and CI in a list

# pksensi 1.1.5

* `solve_mcsim()`: extend the flexible for the model in subfolder 
* Update the vignettes (add deSolve modeling example)

# pksensi 1.1.4

* Replace SE to NSE in `heat_check()`
* Update manual

# pksensi 1.1.3

* Update `solve_fun()` to run deSolve with Pure R code
* set default mxstep = 5000 in `mcsim_install()`

# pksensi 1.1.2

* Add timer in `solve_fun()`
* Add default MCSim `version = '6.1.0'` in `compile_model()`
* Update `pksim()`: Can customize xlab and ylab
* Update `pbtk1cpt model`: Add body weight
* Update vignettes
* Fix bug: Create `.dll` in win10 through `compile_model()`
* Fix bug: `plot()` cannot display variable name when using character
* Revise the `mcsim_version()` for Windows OS

# pksensi 1.1.1

### Fix Bug:

* Add `rtol` and `atol` in `solve_mcsim()`
* Switch default compiler from Clang to GCC in MacOS
* Fix the mName problem under Windows environment in `solve_mcsim()`

### Update vignette

* Remind to set the `mxstep` to 5000 in APAP-PBPK example
* Add note to APAP vignette (Set `atol` and `rtol`)

### Update function

* mcsim_version() can work in MCSim_under_R
* Stop the function when compile error in `compile_model()`
* Revise the default name of output to "simmc.out" and "setpts.out" in `solve_mcsim()`


# pksensi 1.1.0

### Fix Bug:

* Used single time point in `solve_mcsim()`

### Update function:

**Installation**

* Update the default MCSim version to `version = 6.1.0` in `install_mcsim()`
* Adopt the name of model file ("model.R.exe") to MCSim_under_R project

**Modeling**

* Change assignment `n` to `monte_carlo` in `solve_mcsim()`
* Revise the default name of output to "sim.out" and "setpoint.out" in `solve_mcsim()`
* Add solving message to track time spend in `solve_mcsim()` 
* Add assignment `tell = T` to automatically combine the output y in decoupling simulation x in `solve_fun()` and `solve_mcsim()`

**Plot**

* Transfer the log-transformed value to natural scale in `pksim()`
* Revise the discrete time condition to `length(times) < 16` in `heat_check()`


# pksensi 1.0.1

### New vignette:

* Add "PBTK 1-compartment model"
* Add "Acetaminophen-PBPK model"

### New example:

* Update example in `solve_mcsim()`

### New function:

* Add `pbtk1cpt_model()`
* Add `pbpk_apap_model()`
* Add function `mcsim_version()`

### NEW dataset:

* Add `APAP` dataset

### Fix Bug:

- Fix the `order` argument in `heat_check()`
- Add message in `generate_infile()`

### Update function:

* Remove argument `params` in `solve_fun()`
* No need to define `infile.name` and `outfile.name` in `solve_mcsim()`

### Change:

* Change function's name `install_mcsim()` to `mcsim_install()`


# pksensi 1.0.0

* Initial release in CRAN


R/check.R
#' Check the Parameter Sensitivity
#'
#' Visualize and check the sensitivity (or convergence) measurement with the given result.
#'
#' The convergence of sensitivity indices for each parameter is using the approach proposed
#' by Sarrazin et al. (2016).
#' This method quantitatively assesses the convergence by computing the range of
#' 95\% confidence intervals of the sensitivity indices for each parameter across all data points (time and outputs).
#' Using the global approach based on the heatmap visualization combined with the index "cut-off,"
#' can systematically distinguish between "influential" and "non-influential" parameters (Hsieh et al. 2018).
#'
#' @param x a list of the storing information in the defined sensitivity function.
#' @param order a vector of the interested output index, including \code{first order}, \code{interaction}, and \code{total order}.
#' @param vars a logical value or character to specify the display variable in simulation.
#' @param times a logical value or character to specify the display time in simulation.
#' @param cutoff a value or vector to set the cut-off for sensitivity (or convergence) index.
#' The default is 0.05 and 0.1.
#' @param index a character to choose sensitivity index \code{SI} (default) or convergence index \code{CI}.
#' @param level a logical value to use continuous or discrete (default) output.
#' @param text a logical value to display the calculated indices in the plot.
#' @param out a logical value to print the checking result to the console.
#' @param show.all a logical value to show all testing parameters in the heatmap. The default is set to \code{FALSE} to show only the influential parameters.
#' @param ... additional arguments to customize the graphical parameters.
#'
#' @importFrom data.table as.data.table
#' @importFrom stats time
#' @importFrom grDevices colorRampPalette
#' @importFrom graphics barplot legend lines par abline plot.new
#' @importFrom stats runif fft
#' @import ggplot2
#'
#' @return The \code{print} function returns sensitivity and convergence indices
#' with given time-step in the console. The \code{check} method provides the summary of
#' parameter sensitivity and convergence according to the given \code{cutoff}.
#' It can distinguish the influential and non-influential parameter by the providing value
#' of \code{cutoff}. The \code{plot} function can generate the
#' time-course functional outputs of first order and interaction indices for each parameter.
#' The default output is the first model variable. The \code{heat_check} provides a convenient way
#' to visualize and distinguish the influential and non-influential parameter by the setting cut-off.
#' The convergence index can examine the stability of the sensitivity index.
#' To check convergence, be sure to conduct the replication in \code{rfast99}.
#'
#' @references
#' Sarrazin, F., Pianosi, F., & Wagener, T. (2016).
#' Global Sensitivity Analysis of environmental models: Convergence and validation.
#' \emph{Environmental Modelling & Software}, 79, 135–152.
#'
#' Hsieh, N. H., Reisfeld, B., Bois, F. Y., & Chiu, W. A. (2018).
#' Applying a global sensitivity analysis workflow to improve the computational efficiencies
#' in physiologically-based pharmacokinetic modeling.
#' \emph{Frontiers in Pharmacology}, 9, 588.
#'
#' @examples
#' q <- "qunif"
#' q.arg <- list(list(min = 0.6, max = 1),
#'   list(min = 0.5, max = 1.5),
#'   list(min = 0.02, max = 0.3),
#'   list(min = 20, max = 60))
#'
#' params <- c("F","KA","KE","V")
#'
#' set.seed(1234)
#' x <- rfast99(params = params, n = 200, q = q, q.arg = q.arg, rep = 20)
#'
#' time <- c(0.25, 0.5, 1, 2, 4, 6, 8, 12, 24)
#' out <- solve_fun(x, model = FFPK, time = time, vars = "output")
#'
#'
#' # Check results of sensitivity measures
#' check(out)
#' plot(out)
#' heat_check(out, show.all = TRUE)
#' heat_check(out, index = "CI")
#'
#' @seealso \code{\link{tell2}}
#'
#' @rdname check
#' @export
check <- function(x, times, vars, cutoff, out) UseMethod("check")

#' @method check rfast99
#' @export
check.rfast99 <- function(x, times = NULL, vars = NULL, cutoff = 0.05, out = TRUE){

  SI.cutoff <- CI.cutoff <- cutoff

  if (is.null(times)) times <- dimnames(x$y)[[3]] else
    times <- as.character(times)
  if (is.null(vars)) vars <- dimnames(x$y)[[4]]

  if (length(times) == 1 && length(vars) == 1) {

    mSI <- x$mSI[times,,vars]
    iSI <- x$iSI[times,,vars]
    tSI <- x$tSI[times,,vars]
    mCI <- x$mCI[times,,vars]
    iCI <- x$iCI[times,,vars]
    tCI <- x$tCI[times,,vars]

  } else if (length(times) == 1 && length(vars) > 1) {

    mSI <- apply(x$mSI[times,,vars], 1, max)
    iSI <- apply(x$iSI[times,,vars], 1, max)
    tSI <- apply(x$tSI[times,,vars], 1, max)
    mCI <- apply(x$mCI[times,,vars], 1, max)
    iCI <- apply(x$iCI[times,,vars], 1, max)
    tCI <- apply(x$tCI[times,,vars], 1, max)

  } else if (length(times) > 1 | length(vars) > 1) {

    mSI <- apply(x$mSI[times,,vars], 2, max)
    iSI <- apply(x$iSI[times,,vars], 2, max)
    tSI <- apply(x$tSI[times,,vars], 2, max)
    mCI <- apply(x$mCI[times,,vars], 2, max)
    iCI <- apply(x$iCI[times,,vars], 2, max)
    tCI <- apply(x$tCI[times,,vars], 2, max)

  } else {

    mSI <- apply(x$mSI, 2, max)
    iSI <- apply(x$iSI, 2, max)
    tSI <- apply(x$tSI, 2, max)
    mCI <- apply(x$mCI, 2, max)
    iCI <- apply(x$iCI, 2, max)
    tCI <- apply(x$tCI, 2, max)

  }

  # else{
  #    mSI <- x$mSI
  #    iSI <- x$iSI
  #    tSI <- x$tSI
  #    mCI <- x$mCI
  #    iCI <- x$iCI
  #    tCI <- x$tCI
  #  }

  if (out) {
    cat("\nSensitivity check ( Index >", SI.cutoff, ")\n")
    cat("----------------------------------")
    cat("\nFirst order:\n", names(which(mSI > SI.cutoff)), "\n")
    cat("\nInteraction:\n", names(which(iSI > SI.cutoff)), "\n")
    cat("\nTotal order:\n", names(which(tSI > SI.cutoff)), "\n")
    cat("\nUnselected factors in total order:\n", names(which(tSI <= SI.cutoff)), "\n")
    cat("\n")

    cat("\nConvergence check ( Index >", CI.cutoff, ")\n")
    cat("----------------------------------")
    cat("\nFirst order:\n", names(which(mCI > CI.cutoff)), "\n")
    cat("\nInteraction:\n", names(which(iCI > CI.cutoff)), "\n")
    cat("\nTotal order:\n", names(which(tCI > CI.cutoff)), "\n")
    cat("\n")
  }

  x <- list(mSI = names(which(mSI > SI.cutoff)), iSI = names(which(iSI > SI.cutoff)), tSI = names(which(tSI > SI.cutoff)),
            mCI = names(which(mCI > CI.cutoff)), iCI = names(which(iCI > CI.cutoff)), tCI = names(which(tCI > CI.cutoff)))
}

#' @rdname check
#' @export
heat_check <- function(x,
                       order = c("first order", "interaction", "total order"),
                       vars = NULL, times = NULL,
                       index = "SI", cutoff = c(0.05, 0.1),
                       level = T, text = F, show.all = FALSE){

  if (index ==  "SI"){
    SI.cutoff <- cutoff
    nSI <- length(SI.cutoff)
    SI.labels<-rep(NA, nSI+1)

    for(i in 1:nSI){
      SI.labels[i+1] <- paste0(SI.cutoff[i]," - ",SI.cutoff[i+1])
    }
    SI.labels[1] <- paste0("0 - ", SI.cutoff[1])
    SI.labels[nSI+1] <- paste0(" > ", SI.cutoff[nSI])
  }

  if (index ==  "CI"){
    CI.cutoff <- cutoff
    nCI <- length(CI.cutoff)
    CI.labels<-rep(NA, nCI+1)

    for(i in 1:nCI){
      CI.labels[i+1] <- paste0(CI.cutoff[i]," - ",CI.cutoff[i+1])
    }
    CI.labels[1] <- paste0("0 - ", CI.cutoff[1])
    CI.labels[nCI+1] <- paste0(" > ", CI.cutoff[nCI])
  }


  if (index ==  "SI"){
    X_index <- tidy_index(x, index = index)
    X <- cbind(X_index,
               cut(X_index$value, breaks=c(-Inf, paste(SI.cutoff), Inf), labels=SI.labels))
    names(X)[6] <- "level"

    if (!(show.all == TRUE)) {
      check.out <- check.rfast99(x, vars = vars, times, cutoff = min(SI.cutoff), out = F)

      X <- X[X$parameter %in% check.out$tSI, ]

      message(paste0("Display ", length(check.out$tSI),
                     " influential parameters from all ",
                     dim(x$a)[3], " examined parameters."))
    }

  } else if ((index == "CI")) {
    X_index <- tidy_index(x, index = index)
    X <- cbind(X_index,
               cut(X_index$value, breaks=c(-Inf, paste(CI.cutoff), Inf), labels=CI.labels))
    names(X)[6] <- "level"
  }

  colfunc <- colorRampPalette(c("red", "grey90"))

  if (index == "SI"){
    cols <- rev(colfunc(nSI+1))
  } else if (index == "CI") cols <- rev(colfunc(nCI+1))

  X$variable = factor(X$variable, levels=dimnames(x$y)[[4]])
  X$parameter = factor(X$parameter, levels=rev(x$params))

  if (is.null(vars)){
    vars <- dimnames(x$y)[[4]]
  } else (vars <- vars)

  if (is.null(times)){
    times <- dimnames(x$y)[[3]]
  } else (times <- times)

  odr <- order
  variable <- vars
  X <- subset(X, order %in% odr & variable %in% vars & time %in% times)

  #if(length(times) < 16){
  #  X$time <- as.factor(X$time)
  #}
  X$time <- as.factor(X$time)
  X$time <- factor(X$time, levels = times)

  #if (order == F){
  p <- ggplot(X, aes_string("time", "parameter"))
  #} else if (order == T) {
  #  p <- ggplot(X, aes_string("time", "reorder(parameter, value)"))
  #}

  if (level == T) {
    p <- p + geom_tile(aes(fill = level), colour = "white") +
      scale_fill_manual(values = cols)
  } else if (level == F) {
    p <- p + geom_tile(aes_string(fill = "value")) +
      scale_fill_gradient(low = "white", high = "red", limits = c(-0.05,1.05))
  }

  p <- p + scale_x_discrete(expand=c(0,0))
  #if(length(times) < 16) p <- p + scale_x_discrete(expand=c(0,0)) else
  #  p <- p + scale_x_continuous(expand=c(0,0))

  if (length(order) == 1){
    p <- p + scale_y_discrete(expand=c(0,0)) +
      facet_grid(~variable) +
      theme(axis.text.x = element_text(size=10, hjust = 1),
            axis.text.y = element_text(size=10), legend.title=element_blank(),
            legend.position="top")
  } else {
    p <- p + scale_y_discrete(expand=c(0,0)) +
      facet_grid(variable~order) +
      theme(axis.text.x = element_text(size=10, hjust = 1),
            axis.text.y = element_text(size=10), legend.title=element_blank(),
            legend.position="top")
  }

  if (index ==  "SI"){
    p <- p + labs(title="Sensitivity index", x="time", y="parameters")
  } else if ((index == "CI")) {
    p <- p + labs(title="Convergence index", x="time", y="parameters")
  }

  if (text == T){
    if (index ==  "SI"){
      p + geom_text(aes_string(label = "ifelse(value < min(SI.cutoff), '', round(value, 2))"), size = 2.5)
    } else if ((index == "CI")) {
      p + geom_text(aes_string(label = "ifelse(value < min(CI.cutoff), '', round(value, 2))"), size = 2.5)
    }
  } else p
}

tidy_index <- function (x, index = "SI") {

  if(index == "CI") {
    m <- as.data.table(x$mCI) |> cbind(order = "first order")
    i <- as.data.table(x$iCI) |> cbind(order = "interaction")
    t <- as.data.table(x$tCI) |> cbind(order = "total order")
    X <- do.call(rbind, list(m, i, t))
  } else if (index == "SI") {
    m <- as.data.table(x$mSI) |> cbind(order = "first order")
    i <- as.data.table(x$iSI) |> cbind(order = "interaction")
    t <- as.data.table(x$tSI) |> cbind(order = "total order")
    X <- do.call(rbind, list(m, i, t))
  }
  names(X) <- c("time", "parameter", "variable", "value", "order")
  return(X)
}

#' @rdname check
#' @method plot rfast99
#' @export
plot.rfast99 <- function(x, vars = 1, cutoff = 0.1, ...){

  mSI <- x$mSI[,,vars]
  iSI <- x$tSI[,,vars]
  tSI <- x$tSI[,,vars]
  mCI <- x$mCI[,,vars]
  iCI <- x$tCI[,,vars]
  tCI <- x$tCI[,,vars]

  if(is.matrix(mSI)){

    nv <- length(colnames(tSI))+1
    nc <- ceiling(sqrt(nv))
    nr <- ceiling(nv/nc)

    times <- row.names(tSI)

    old.par <- par(no.readonly = TRUE)
    par(mfrow = c(nr, nc), mar = c(4,2,3,1), oma = c(0,0,2,0))

    for(i in 1:ncol(tSI)){
      plot(times, tSI[,i], ylim = c(0, 1), bty = 'n',
           type = 'l', lwd = 2, xlab = 'time', ylab = '',
           main = colnames(tSI)[i], ...)
      col.transp = adjustcolor('black', alpha.f = 0.4)
      polygon(x = c(times, rev(times)),
              y =c(tSI[,i]-tCI[,i], rev(tSI[,i]+tCI[,i])),
              col = col.transp, border = col.transp)

      col.transp = adjustcolor('red', alpha.f = 0.4)
      lines(times, mSI[,i], ylim = c(0, 1), bty = 'n',
            lwd = 2, col = 'red')
      polygon(x = c(times, rev(times)),
              y =c(mSI[,i]-mCI[,i], rev(mSI[,i]+mCI[,i])),
              col = col.transp, border = col.transp)
      if (is.numeric(cutoff)){
        abline(cutoff, 0, lty = 2)
      }
    }

    if (is.character(vars)) {vars <- which(dimnames(x$y)[[4]] == vars)}

    variable <- dimnames(x$y)[[4]][vars]
    mtext(variable, NORTH<-3, line=0.4, adj=0, cex=1.5, outer=TRUE)

    plot.new()
    legend('top', legend = c('total order', 'first order'), col = c('black','red'),
           lty = 1, lwd = 1, pch = NA, bty = 'n',
           text.col = 'black')
    par(old.par)
  }
}

#' @rdname check
#' @method print rfast99
#' @export
print.rfast99 <- function(x, ...) {
  cat("\nCall:\n", deparse(x$call), "\n", sep = "")
  if (! is.null(x$y) && ! is.null(x$mSI)) {
    digits <- 4

    cat("\nModel runs:", dim(x$y)[1], "\n")
    cat("\n")
    cat("\n==================================")
    cat("\nSensitivity Indices", "\n")
    cat("\nfirst order:", "\n")
    print(round(x$mSI, digits = digits))
    cat("\ninteraction:", "\n")
    print(round(x$iSI, digits = digits))
    cat("\ntotal order:", "\n")
    print(round(x$tSI, digits = digits))

    if (x$rep > 1){ # w/ replication
      cat("\n")
      cat("\n=================================")
      cat("\nConvergence Indices", "\n")
      cat("\nfirst order:", "\n")
      print(round(x$mCI, digits = digits))
      cat("\ninteraction:", "\n")
      print(round(x$iCI, digits = digits))
      cat("\ntotal order:", "\n")
      print(round(x$tCI, digits = digits))
    }
  }
  else {
    cat("(empty)\n")
  }
}
